### PR TITLE
Add @top_level to access the top-level scope in macros

### LINKED ISSUE #8914: Add top-level has_constant? macro method (or expose Program type)
In #1585 `has_constant?` was added and allows you to check if a constant is defined within a type. But I think there is no hay to check if a top-level constant exists.

Adding a top-level macro method to check if the constant is defined in the internal `Program` type that holds the top level could look like:

```crystal
{{ has_constant?(:MyShard) }}
```

Or if we expose the top level as an alias we could directly use.

```crystal
{{ Program.has_constant?(:MyShard) }}
```

It can be useful to check if a shard is required or not. And to use feature checks instead of comparing versions.

### PR DESCRIPTION
**Pull request by beta-ziliani**
Fixes #8914 by creating a new instance method `@top_level` (as suggested by @jhass). This allows to query the top level module, as the following example shows:

```crystal
FOO = 1
class Foo
   def bar
       {{@top_level.has_constant?("FOO")}}
    end
end
```

### CODE REVIEW

File: src/compiler/crystal/macros/interpreter.cr [hunk 1]
```diff
@@ -546,6 +546,8 @@ module Crystal
       when "@type"
         target = @scope == @program.class_type ? @scope : @scope.instance_type
         @last = TypeNode.new(target.devirtualize)
+      when "@top_level"
+        @last = TypeNode.new(@program.devirtualize)
```
Comment (asterite) [inline src/compiler/crystal/macros/interpreter.cr:550]: `devirtualize` isn't needed here, but using it does no harm (it just returns `self` in the case of `@program`)

```suggestion
        @last = TypeNode.new(@program)
```

**Review by asterite — APPROVED** [review]
Looks great, thank you!

File: spec/compiler/codegen/macro_spec.cr [hunk 2]
```diff
@@ -1876,4 +1876,29 @@ describe "Code gen: macro" do
...
+  it "access to the program with @top_level" do
+    run(%(
+      class Foo
+        def bar
+          {{@top_level.name.stringify}}
+        end
+      end
+
+      Foo.new.bar
+      )).to_string.should eq("main")
```
Comment (straight-shoota) [inline spec/compiler/codegen/macro_spec.cr:1889]: This is not directly related, but it feels a bit odd that the top-level scope is called `main`. `top_level` would probably be more appropriate. But I', not sure what it should be. 
`@top_level.stringify` is `<Program>`, btw.

Comment (beta-ziliani) [reply spec/compiler/codegen/macro_spec.cr:1889]: top_level sounds good. TBH this was a test "after the fact", and I felt that naming it differently was another issue. Oddly enough, I couldn't find where that comes from...

Comment (asterite) [reply spec/compiler/codegen/macro_spec.cr:1889]: https://github.com/crystal-lang/crystal/blob/421fa11bf05b11441b74bbeb2762b673f96fec3f/src/compiler/crystal/program.cr#L126

It should be fine to change it, it was never shown until now.

Comment (beta-ziliani) [reply spec/compiler/codegen/macro_spec.cr]: I took the liberty to add the change in this PR. Thanks @asterite !

**Review by straight-shoota — APPROVED** [review]

**Comment by straight-shoota** [comment]
Gracias, @beta-ziliani
